src/System.ComponentModel.EventBasedAsync/tests/BackgroundWorkerTests.cs

```diff
@@ -214,6 +214,46 @@ public void RunWorkerAsyncTwice()
         }
 
         [Fact]
+        public void TestCancelInsideDoWork()
+        {
+            var original = SynchronizationContext.Current;
+            try
+            {
+                SynchronizationContext.SetSynchronizationContext(null);
+
+                var bw = new BackgroundWorker() { WorkerSupportsCancellation = true };
+                var barrier = new Barrier(2);
+
+                bw.DoWork += (sender, e) =>
+                {
+                    barrier.SignalAndWait();
+                    barrier.SignalAndWait();
+
+                    if (bw.CancellationPending)
+                    {
+                        e.Cancel = true;
+                    }
+                };
+
+                bw.RunWorkerCompleted += (sender, e) =>
+                {
+                    Assert.True(e.Cancelled);
+                    barrier.SignalAndWait();
+                };
+
+                bw.RunWorkerAsync();
+                barrier.SignalAndWait();
+                bw.CancelAsync();
+                barrier.SignalAndWait();
+                Assert.True(barrier.SignalAndWait(TimeoutShort), "Background work timeout");
```

Comment:
Should this be TimeoutLong rather than TimeoutShort?  I believe TimeoutLong is typically used where we expect the work to always complete and we require waiting until it's done, but we want to bail if it takes so long that there must be an issue... by using TimeoutShort here, I think it's likely now and then we'd hit a spurious failure.